### PR TITLE
Support passing loss func in compile Scala

### PR DESCRIPTION
**Pull request by hkvision**
Fix https://github.com/intel-analytics/zoo/issues/174

### CODE REVIEW

**Comment by hkvision** [comment]
Using CustomLossWithFunc will raise the following exception cc @zhichao-li 
```
org.scalatest.Assertions$AssertionsHelper
Serialization stack:
	- object not serializable (class: org.scalatest.Assertions$AssertionsHelper, value: org.scalatest.Assertions$AssertionsHelper@5482bcac)
	- field (class: org.scalatest.FlatSpec, name: assertionsHelper, type: class org.scalatest.Assertions$AssertionsHelper)
	- object (class com.intel.analytics.zoo.pipeline.api.keras.models.TrainingSpec, TrainingSpec)
	- field (class: com.intel.analytics.zoo.pipeline.api.keras.models.TrainingSpec$$anonfun$2, name: $outer, type: class com.intel.analytics.zoo.pipeline.api.keras.models.TrainingSpec)
	- object (class com.intel.analytics.zoo.pipeline.api.keras.models.TrainingSpec$$anonfun$2, <function0>)
	- field (class: com.intel.analytics.zoo.pipeline.api.keras.models.TrainingSpec$$anonfun$2$$anonfun$7, name: $outer, type: class com.intel.analytics.zoo.pipeline.api.keras.models.TrainingSpec$$anonfun$2)
	- object (class com.intel.analytics.zoo.pipeline.api.keras.models.TrainingSpec$$anonfun$2$$anonfun$7, <function2>)
	- field (class: com.intel.analytics.zoo.pipeline.api.autograd.CustomLossWithFunc, name: func, type: interface scala.Function2)
	- object (class com.intel.analytics.zoo.pipeline.api.autograd.CustomLossWithFunc, com.intel.analytics.zoo.pipeline.api.autograd.CustomLossWithFunc@0)
	- field (class: scala.Tuple4, name: _1, type: class java.lang.Object)
	- object (class scala.Tuple4, (com.intel.analytics.zoo.pipeline.api.autograd.CustomLossWithFunc@0, {
	maxDropPercentage: 0.0
	computeThresholdbatchSize: 100
	warmupIterationNum: 200
	isLayerwiseScaled: false
	dropPercentage: 0.0
 },None,com.intel.analytics.bigdl.optim.SGD$mcF$sp@45e6d1e0))
java.io.NotSerializableException: org.scalatest.Assertions$AssertionsHelper
```

**Comment by zhichao-li** [comment]
Need to dig this into details after the workshop, or I think we can even just delete this class as python is using CustomLossWithVariable

**Comment by hkvision** [comment]
Using CustomLossWithVariable can work (just as how Python side works), see the unit test I add.

**Comment by hkvision** [comment]
cc @jason-dai

**Comment by jason-dai** [comment]
LGTM

**Comment by jason-dai** [comment]
Is this ready for merge?

**Comment by hkvision** [comment]
Almost ready. Will merge this by today. Need some discussion whether `CustomLossWithFunc` will be removed or investigation why `CustomLossWithFunc` will have serialization issue.

**Comment by zhichao-li** [comment]
Let's simply private `CustomLossWithFunc` and merge this once pass the Jenkins test.
We can refactor `CustomLossWithFunc` in a following PR.

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/job/ZOO-PR-Validation/368/
